stumpy/mmotifs.py

```diff
@@ -0,0 +1,187 @@
+        If None, this defaults to
+        `np.nanmax([np.nanmean(D) - 2 * np.nanstd(D), np.nanmin(D)])`
+        (i.e. at least the closest match will be returned).
+
+    cutoff: numpy.ndarray or float, default None
+        The largest matrix profile value (distance) for each dimension of the
+        multidimensional matrix profile that a multidimenisonal candidate motif is
+        allowed to have.
+        If cutoff is only one value, these value will be applied to every dimension.
+
+    max_matches: int, default 10
+        The maximum amount of similar matches (nearest neighbors) of a motif
+        representative to be returned.
+        The first match is always the self-match for each motif.
+
+    max_motifs: int, default 1
+        The maximum number of motifs to return
+
+    atol : float, default 1e-8
+        The absolute tolerance parameter. This value will be added to `max_distance`
+        when comparing distances between subsequences.
+
+    Returns
+    -------
+    motif_matches_distances: numpy.ndarray
+        The distances corresponding to a set of subsequence matches for each motif.
+
+    motif_matches_indices: numpy.ndarray
+        The indices corresponding to a set of subsequences matches for each motif.
+
+    motif_subspaces: numpy.ndarray
+        A numpy.ndarray consisting of arrays that contain the `k`-dimensional
+        subspace for each motif.
+
+    motif_mdls: numpy.ndarray
+        A numpy.ndarray consisting of arrays that contain the mdl results for
+        finding the dimension of each motif
+    """
+    if max_motifs < 1:  # pragma: no cover
+        logger.warning(
+            "The maximum number of motifs, `max_motifs`, "
+            "must be greater than or equal to 1"
+        )
+        logger.warning("`max_motifs` has been set to `1`")
+        max_motifs = 1
+
+    # Calculate subsequence length
+    T = core._preprocess(T)  # convert dataframe to array if necessary
+    m = T.shape[-1] - P.shape[-1] + 1
+
+    # Calculate exclusion zone
+    excl_zone = int(np.ceil(m / config.STUMPY_EXCL_ZONE_DENOM))
+
+    # Precompute rolling means and standard deviations
+    T, mean_T, sigma_T = core.preprocess(T, m)
+
+    if max_matches is None:
+        max_matches = np.inf
+
+    motif_matches_distances = []
+    motif_matches_indices = []
+    motif_subspaces = []
+    motif_mdls = []
+
+    # Identify the first multidimensional motif as candidate motif and get its
+    # nearest neighbor
+    candidate_idx = np.argsort(P, axis=1)[:, 0]
+    nn_idx = I[np.arange(len(candidate_idx)), candidate_idx]
+
+    # Iterate over each motif
+    while len(motif_matches_distances) < max_motifs:
+
+        # Choose dimension k using MDL
+        mdls, subspaces = mdl(T, m, candidate_idx, nn_idx)
+        k = np.argmin(mdls)
+        motif_mdls.append(mdls)
+
+        # Calculate default cutoff value
+        if cutoff is None:

```

Comment:
Maybe I'm missing something but I think that this entire `cutoff` section should not be re-computed during each iteration of the `while` loop. I think this should be moved outside of the `while` loop and the cutoffs should be computed once for each dimension and that is it.

Reply:
I don't know. I only computed the cutoff value for the `k`-th dimension since I thought that it is likely that one has more dimensions than motifs. We only need the `cutoff` value for the `k`-th dimension, am I right? Since we have to compare `motif_value` with `cutoff_var` later and `motif_value` is only for the `k`-th dimension (`motif_value = P[k, motif_idx]`). Furthermore I check whether `cutoff` is `None`, an array/list or a single value. If it is a list/array I also take only the `k`-th value since this belongs to the `k`-th dimension of the matrix profile and if it is a single value this value should be applied to every iteration (therefore for every dimension, that is needed). Shouldn't this be easier / computationally favorable for most cases than computing the `cutoff` value for each dimension at the beginning and taking out the `k`-th value on each iteration of the while-loop?

Reply:
All right, I have tried to do it as you suggested and moved the `cutoff` section outside of the `while` loop. I think it should work the same now but the code could certainly be improved. I am totally unexperienced when it comes to runtime optimization or code aesthetics 😄